src/interface/metadata.hpp

```diff
@@ -166,67 +166,65 @@ class Metadata {
   PARTHENON_INTERNAL_FOREACH_BUILTIN_FLAG
 #undef PARTHENON_INTERNAL_FOR_FLAG
 
-  /// Default constructor override
-  Metadata() : shape_({1}), sparse_id_(-1) {}
-
-  /// returns a new Metadata instance with set bits,
-  /// set sparse_id, and fourth dimension
-  explicit Metadata(const std::vector<MetadataFlag> &bits) : shape_({1}), sparse_id_(-1) {
-    SetMultiple(bits);
-  }
-
-  /// returns a metadata with bits and shape set
-  explicit Metadata(const std::vector<MetadataFlag> &bits, const std::vector<int> &shape)
-      : shape_(shape), sparse_id_(-1) {
-    SetMultiple(bits);
-  }
-
-  /// returns a metadata with bits and shape set
-  explicit Metadata(const std::vector<MetadataFlag> &bits, const std::vector<int> &shape,
-                    const std::vector<std::string> component_labels)
-      : shape_(shape), sparse_id_(-1), component_labels_(component_labels) {
-    SetMultiple(bits);
-  }
-
-  /// returns a metadata with bits and sparse id set
-  explicit Metadata(const std::vector<MetadataFlag> &bits, const int sparse_id)
-      : shape_({1}), sparse_id_(sparse_id) {
-    SetMultiple(bits);
-    PARTHENON_REQUIRE_THROWS(IsSet(Sparse), "Sparse ID requires sparse metadata");
-  }
-
-  explicit Metadata(const std::vector<MetadataFlag> &bits, const std::string &associated)
-      : associated_(associated) {
-    SetMultiple(bits);
-  }
-
-  /// returns a metadata with bits, shape, and sparse ID set
-  explicit Metadata(const std::vector<MetadataFlag> &bits, int sparse_id,
-                    const std::vector<int> &shape)
-      : shape_(shape), sparse_id_(sparse_id) {
-    SetMultiple(bits);
-    PARTHENON_REQUIRE_THROWS(IsSet(Sparse), "Sparse ID requires sparse metadata");
-  }
-
-  explicit Metadata(const std::vector<MetadataFlag> &bits, const int sparse_id,
-                    const std::string &associated)
-      : sparse_id_(sparse_id), associated_(associated) {
-    SetMultiple(bits);
-    PARTHENON_REQUIRE_THROWS(IsSet(Sparse), "Sparse ID requires sparse metadata");
-  }
-
-  explicit Metadata(const std::vector<MetadataFlag> &bits, const std::string &associated,
-                    const std::vector<int> &shape)
-      : associated_(associated), shape_(shape) {
+  Metadata() = default;
+
+  // There are 4 optional arguments: shape, sparse_id, component_labels, and associated,
+  // so we'll need 16 constructors to provide all possible variants
+
+  // 5 constructors, this is the general constructor called by all other constructors, so
+  // we do some sanity checks here
+  Metadata(const std::vector<MetadataFlag> &bits, const std::vector<int> &shape = {1},
+           int sparse_id = -1, const std::vector<std::string> component_labels = {},
```

Comment:
I think you can make component_labels a reference to avoid a copy.

Reply:
Good catch, this should be a ref. Thanks!